multisense_ros/src/ground_surface_utilities.cpp

```diff
@@ -0,0 +1,279 @@
+        }
+    }
+    return functionValue;
+}
+
+} // anonymous namespace
+
+Eigen::Matrix<uint8_t, 3, 1> groundSurfaceClassToPixelColor(const uint8_t value)
+{
+    if (value == 1) // Out of bounds
+        return Eigen::Matrix<uint8_t, 3, 1>{ 114, 159, 207 };
+    else if (value == 2) // Obstacle
+        return Eigen::Matrix<uint8_t, 3, 1>{ 255, 0, 0 };
+    else if (value == 3) // Free space
+        return Eigen::Matrix<uint8_t, 3, 1>{ 0, 255, 0 };
+
+    // Unknown
+    return Eigen::Matrix<uint8_t, 3, 1>{ 0, 0, 0 };
+}
+
+sensor_msgs::PointCloud2 eigenToPointcloud(
+    const std::vector<Eigen::Vector3f> &input,
+    const std::string &frame_id)
+{
+    sensor_msgs::PointCloud2 ret =
+        multisense_ros::initialize_pointcloud<float>(true, frame_id, "intensity");
+
+    const double num_points = input.size();
+    ret.data.resize(num_points * ret.point_step);
+
+    for (size_t i = 0; i < num_points; ++i)
+    {
+        float* cloudP = reinterpret_cast<float*>(&(ret.data[i * ret.point_step]));
+        cloudP[0] = input[i][0];
+        cloudP[1] = input[i][1];
+        cloudP[2] = input[i][2];
+    }
+
+    ret.height = 1;
+    ret.row_step = num_points * ret.point_step;
+    ret.width = num_points;
+    ret.data.resize(num_points * ret.point_step);
+
+    return ret;
+}
+
+std::vector<Eigen::Vector3f> convertSplineToPointcloud(
+    const Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> &controlGrid,
+    const float* xzCellOrigin,
+    const float* xzCellSize,
+    const float* xzLimit,
+    const float* minMaxAzimuthAngle,
+    const float* extrinsics,
+    const float* quadraticParams)
+{
+    static constexpr double drawResolution = 0.1;
+    static const auto basisArray = generateBasisArray();
+
+    // Generate extrinsics matrix
+    Eigen::Matrix<float, 4, 4> extrinsicsMat;
+    {
+        extrinsicsMat.setZero();
+        extrinsicsMat(0, 3) = extrinsics[0];
+        extrinsicsMat(1, 3) = extrinsics[1];
+        extrinsicsMat(2, 3) = extrinsics[2];
+        extrinsicsMat(3, 3) = static_cast<float>(1.0);
+        Eigen::Matrix<float, 3, 3> rot =
+            (Eigen::AngleAxis<float>(extrinsics[5], Eigen::Matrix<float, 3, 1>(0, 0, 1))
+            * Eigen::AngleAxis<float>(extrinsics[4], Eigen::Matrix<float, 3, 1>(0, 1, 0))
+            * Eigen::AngleAxis<float>(extrinsics[3], Eigen::Matrix<float, 3, 1>(1, 0, 0))).matrix();
+        extrinsicsMat.block(0, 0, 3, 3) = rot;
+    }
+
+    // Precompute extrinsics inverse
+    const auto extrinsicsInverse = extrinsicsMat.inverse();
+
+    // Get boundaries of valid spline area
+    const auto minX = xzCellOrigin[0] + xzCellSize[0];
+    const auto minZ = xzCellOrigin[1] + xzCellSize[1];
+    const auto maxX = xzLimit[0];
+    const auto maxZ = xzLimit[1];
+
+    // Precompute number of points
+    size_t numPoints = 0;
+    for (float x = minX; x < maxX; x += drawResolution)
+        for (float y = minZ; y < maxZ; y += drawResolution)
+            numPoints++;
```

Comment:
Can you do this without two for loops? Wouldn't something like this work?

`const size_t numPoints = std::floor((maxX - minX) / drawResolution) * std::floor((maxY - minY) / drawResolution)`

Reply:
Absolutely, thanks for picking up on this (lazy programming on my behalf)!